### PR TITLE
[caffe2][nomnigraph] Add binary split algorithm to Algorithms.h

### PR DESCRIPTION
**Pull request by bwasti**
This diff will allow refactoring some of the current backend_cut code

### CODE REVIEW

File: caffe2/core/nomnigraph/include/nomnigraph/Graph/Algorithms.h [hunk 1]
```diff
@@ -26,6 +26,87 @@ template <typename T, typename U>
...
+  // Topologically sorted nodes.
+  std::vector<Subgraph<T,U>> sortedNodes;
+  sortedNodes.emplace_back();
+
+  // Find the initial frontier.
+  std::vector<typename G::NodeRef> frontier;
+  std::vector<typename G::NodeRef> nextFrontier;
+
+  for (auto n : g->getMutableNodes()) {
+    if (n->getInEdges().size() == 0 && condition(n)) {
```
Comment (yinghai) [inline caffe2/core/nomnigraph/include/nomnigraph/Graph/Algorithms.h]: If no primary input nodes matches condition, the algorithm will fail to start. We probably need to do one iteration to jump start in this case.

File: caffe2/core/nomnigraph/include/nomnigraph/Graph/Algorithms.h [hunk 2]
```diff
@@ -26,6 +26,87 @@ template <typename T, typename U>
...
+      frontier.emplace_back(n);
+    }
+  }
+
+  auto stillHasInEdge = [&](typename G::NodeRef m) {
```
Comment (yinghai) [inline caffe2/core/nomnigraph/include/nomnigraph/Graph/Algorithms.h]: This is not very important but using edgeset is not very efficient. I like the way you did for Tarjan's, where you attach an annotation property with the graph node.

File: caffe2/core/nomnigraph/include/nomnigraph/Graph/Algorithms.h (hunk 1 above)
Comment (bwasti) [reply caffe2/core/nomnigraph/include/nomnigraph/Graph/Algorithms.h]: ah good catch, I think it'd be as simple as adding the inverse of the condition to the nextFrontier and then refactoring the while loop

Comment (yinghai) [reply caffe2/core/nomnigraph/include/nomnigraph/Graph/Algorithms.h]: Yes.

File: caffe2/core/nomnigraph/CMakeLists.txt [hunk 3]
```diff
@@ -9,3 +9,15 @@ add_dependencies(nomnigraph Caffe_PROTO Caffe2_PROTO)
 target_include_directories(nomnigraph PUBLIC ${CMAKE_CURRENT_SOURCE_DIR}/include)
 list(APPEND Caffe2_DEPENDENCY_LIBS nomnigraph)
 set(Caffe2_DEPENDENCY_LIBS ${Caffe2_DEPENDENCY_LIBS} PARENT_SCOPE)
+
+
+if (BUILD_TEST)
+  foreach(test_src ${NOMNI_TEST_SRCS})
+    message(INFO ${test_src} "WWWW")
```
Comment (yinghai) [inline caffe2/core/nomnigraph/CMakeLists.txt]: Remove?

File: caffe2/core/nomnigraph/include/nomnigraph/Graph/BinaryMatchImpl.h [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+          frontier.emplace_back(m);
+        } else {
+          nextFrontier.emplace_back(m);
+        }
+      }
+    }
+  }
+
+  if (edgeSet.size()) {
+    assert(0 && "Invalid graph for Kahn's algorithm, cycle detected.  Please use Tarjans.");
```
Comment (yinghai) [inline caffe2/core/nomnigraph/include/nomnigraph/Graph/BinaryMatchImpl.h]: Can we throw instead of assert?

**Review by yinghai — APPROVED** [review]
Looks good! Thanks.

File: caffe2/core/nomnigraph/include/nomnigraph/Graph/BinaryMatchImpl.h (hunk 4 above)
Comment (bwasti) [reply caffe2/core/nomnigraph/include/nomnigraph/Graph/BinaryMatchImpl.h]: I don't have any exceptions in the library but this concern is pretty important.  I'm not sure the best way to handle expected errors while also avoiding exceptions (which was a choice made a while back for codesize reduction and to be consistent with LLVM)

Comment (yinghai) [reply caffe2/core/nomnigraph/include/nomnigraph/Graph/BinaryMatchImpl.h]: Yeah, before at application level, this is not a hard error. Worst case is that we don't optimize the net.

**Pull request merged**